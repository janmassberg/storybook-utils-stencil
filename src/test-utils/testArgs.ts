export const testArgs = {
    testNull: null,
    testUndefined: undefined,
    testFalse: false,
    testTrue: true,
    testString: "test-string",
    testNumber: 12345,
    testArray: [1, 2, 3],
    testObject: {
        foo: "bar",
    },
    testFunction: (): void => {
        // Implementation
    },
    onTestEvent: (event: Event): void => {
        // Handle event
        console.log(event.type);
    },
    _slot: "Test children",
};

export const testArgsButton = {
    required: true,
    variant: "primary",
    onClick: (event: Event) => {
        // Handle click event
        console.log(event.target);
    },
    _slot: "Label",
};

export const testArgsComponentGenerator = {
    component: "my-component",
    args: testArgs,
};

export const testCopmonentSlot = `<my-child-component foo-bar="foo">
<my-nested-child-component bar-foo="bar"><span class="test-class">Label</span></my-nested-child-component>
</my-child-component>`;
