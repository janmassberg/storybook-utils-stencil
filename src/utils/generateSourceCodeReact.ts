import { formatJsx } from "./formatJsx";
import {
    ArgsType,
    filterEventHandlers,
    filterHtmlAttributes,
    filterJsxProperties,
} from "./argUtils";
import { objectToString } from "./objectToString";
import { objectToStringJsx } from "./objectToStringJsx";
import { toLowerCamelCase, toUpperCamelCase, ucFirst } from "./stringUtils";

const transformReactAttributes = (
    attributes: ArgsType
) => {
    return attributes
        .map(([key, value]) => {
            key = toLowerCamelCase(key);
            if (typeof value === "boolean") {
                return value ? key : "";
            }
            if (key === "class") {
                key = "className";
            }
            return `${key}=${objectToStringJsx(value)}`;
        })
        .filter((str) => !!str);
};

const transformReactAttributesString = (str: string): string => {
    return str.replace(
        new RegExp(`([a-z]-[a-z-]+)(="[^"]+")`, "g"),
        (_match, attrib, value) => {
            const valueTransformed = value === '="true"' ? "" : value;
            return `${toLowerCamelCase(attrib)}${valueTransformed}`;
        }
    );
};

const transformChildComponents = (str: string): string => {
    if (!str) {
        return "";
    }
    const matches = str.match(/<([a-z-]+-[a-z-]+)/g);
    const tagNames: string[] = [];
    if (matches && matches.length) {
        matches.forEach((match) => {
            const tagName = match.replace(/[<>/]/g, "");
            if (tagNames.indexOf(tagName) === -1) {
                tagNames.push(tagName);
            }
        });
    }
    tagNames.sort((a, b) => b.length - a.length);
    tagNames.forEach((tagName) => {
        const reactTagName = toUpperCamelCase(tagName);
        str = str.replace(new RegExp(tagName, "g"), reactTagName);
        str = str.replace(
            new RegExp(`<${reactTagName}([^/>]+)`, "g"),
            (_, attributes) => {
                return `<${reactTagName}${transformReactAttributesString(
                    attributes
                )}`;
            }
        );
    });
    return str.replace(/(\s)class=/gi, "$1className=");
};

export const generateSourceCodeReact = (
    component: string,
    args: Record<string, unknown>
) => {
    const attribs = filterHtmlAttributes(args);
    const props = filterJsxProperties(args);
    const events = filterEventHandlers(args);
    const slot = typeof args._children === "string" ? args._children : "";

    const reactComponentName = toUpperCamelCase(component);
    const reactAttributes = transformReactAttributes(attribs);
    const reactProps = props.map(
        ([key]) =>
            `${toLowerCamelCase(key)}={${toLowerCamelCase(key)}}\n`
    );
    const reactEvents = events.map(
        ([key]) => `on${ucFirst(key)}={on${ucFirst(key)}}\n`
    );
    const reactComponentJsx = `<${reactComponentName}
        ${reactAttributes.join(" ")}
        ${reactProps.join(" ")}
        ${reactEvents.join(" ")}
        >${transformChildComponents(slot)}</${reactComponentName}>`;

    const reactEventHandlers = events.map(
        ([key, value]) => `const on${ucFirst(key)} = (event: Event) => {
            ${value
            .toString()
            .replace(/^[^{]+{/, "")
            .replace(/}$/, "")}
        };`
    );

    const reactConstants = props.map(([key, value]) => {
        return `const ${key} = ${objectToString(value)};`;
    });

    const componentName = `${toUpperCamelCase(component)}Example`;
    const reactComponentSource = `
    const ${componentName} = () => {

        ${reactConstants.join("\n\n")}

        ${reactEventHandlers.join("\n\n")}

        return (
            ${reactComponentJsx}
        );
    };
    `;
    return formatJsx(reactComponentSource);
};

