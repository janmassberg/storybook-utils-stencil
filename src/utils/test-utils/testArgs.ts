export const testArgs = {
    testNull: null,
    testUndefined: undefined,
    testFalse: false,
    testTrue: true,
    testString: "test-string",
    testNumber: 12345,
    testArray: [1, 2, 3],
    testObject: {
        foo: "bar",
    },
    testFunction: () => {},
    onTestEvent: () => {},
    _children: "Test children",
};
