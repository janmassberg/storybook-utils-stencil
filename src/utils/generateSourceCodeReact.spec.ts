import { generateSourceCodeReact } from "./generateSourceCodeReact";
import { testArgs } from "../test-utils/testArgs";

describe("generateSourceCodeReact", () => {
    it("should generate valid and formatted react JSX code", () => {
        expect(
            generateSourceCodeReact("my-component", testArgs)
        ).toMatchSnapshot();
    });
});
