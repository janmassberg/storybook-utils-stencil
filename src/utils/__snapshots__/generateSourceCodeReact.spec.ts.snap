// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateSourceCodeReact should generate valid formatted react JSX code with children 1`] = `
"const MyComponentExample = () => {
  return (
    <MyComponent>
      <MyChildComponent fooBar=\\"foo\\">
        <MyNestedChildComponent barFoo=\\"bar\\">
          <span =\\"test-class\\">Test children</span>
        </MyNestedChildComponent>
      </MyChildComponent>
    </MyComponent>
  );
};
"
`;

exports[`generateSourceCodeReact should generate valid formatted react JSX code with props 1`] = `
"const MyComponentExample = () => {
  const testArray = [1, 2, 3];

  const testObject = { foo: \\"bar\\" };

  const testFunction = function () {
    // do nothing
  };

  const onOnTestEvent = (event: Event) => {
    console.log(event.type);
  };

  return (
    <MyComponent
      testTrue
      testString=\\"test-string\\"
      testNumber={12345}
      testArray={testArray}
      testObject={testObject}
      testFunction={testFunction}
      onOnTestEvent={onOnTestEvent}
    >
      Test children
    </MyComponent>
  );
};
"
`;
